tools/lib/provision.py

```diff
@@ -336,20 +336,29 @@ def main(options):
             print("RabbitMQ is already configured.")
 
         migration_status_path = os.path.join(UUID_VAR_PATH, "migration_status_dev")
-        if options.is_force or not is_template_database_current(
-                migration_status=migration_status_path,
-                settings="zproject.settings",
-                database_name="zulip",
-        ):
+        dev_template_db_status = template_database_status(
+            migration_status=migration_status_path,
+            settings="zproject.settings",
+            database_name="zulip",
+        )
+        if options.is_force or dev_template_db_status == 'needs_rebuild':
             run(["tools/setup/postgres-init-dev-db"])
             run(["tools/do-destroy-rebuild-database"])
-        else:
+        elif dev_template_db_status == 'run_migrations':
+            run_db_migrations(
+                migration_status_file="migration_status_dev",
+                settings="zproject.settings",
+            )
+        elif dev_template_db_status == 'current':
             print("No need to regenerate the dev DB.")
 
-        if options.is_force or not is_template_database_current():
+        test_template_db_status = template_database_status()
+        if options.is_force or test_template_db_status == 'needs_rebuild':
             run(["tools/setup/postgres-init-test-db"])
             run(["tools/do-destroy-rebuild-test-database"])
-        else:
+        elif test_template_db_status == 'run_migrations':
+            run_db_migrations()
```

Comment:
I'd prefer that this get an option specifying which one it's doing.

Reply:
like specifically say that it is doing a test_migration rebuild?
I think that can be done. basically I am still fiddling with the migrations on test database. I can't understand why running migrations for dev also seems to migrate test db

Reply:
It's important to understand that there are 3 databases:
* The `zulip` dev database
* The `zulip_test_template` database, which `provision` updates
* The `zulip_test` database, which is generated from the `zulip_test_template` every time you run `test_backend` by copying the template; that way we always start the tests with a fresh state.

Reply:
so basically we want migrations to run on the first two because third doesn't exist yet.
That should mean we need to run migrations on both the db and there should be same migrations pending when that is done. Thing is lets say I just introduce a new variable in userprofile to create a new migration. When I provision after running the migrations during dev db, when time comes to do it for test db the execution of migrate command yields no migration to apply!

Reply:
Here's a traceback, take a look
```
generate_secrets: No new secrets to generate.
+ tools/update-authors-json --use-fixture
+ tools/inline-email-css
RabbitMQ is already configured.
zproject.settings
Operations to perform:
  Apply all migrations: analytics, auth, confirmation, contenttypes, otp_static, otp_totp, pgroonga, sessions, sites, social_django, two_factor, zerver, zilencer
Running migrations:
  Applying zerver.0178_userprofile_long_rm_idle_haeaadadgdada... OK
2018-06-05 14:15:23.189 INFO [] Clearing memcached cache after migrations
zproject.settings
zproject.test_settings
Operations to perform:
  Apply all migrations: analytics, auth, confirmation, contenttypes, otp_static, otp_totp, pgroonga, sessions, sites, social_django, two_factor, zerver, zilencer
Running migrations:
  No migrations to apply.
2018-06-05 14:15:23.717 INFO [] Clearing memcached cache after migrations
zproject.test_settings
+ touch /srv/zulip/var/54db07ed-532f-40bb-acbe-1983e2d30eb9/last_compilemessages_hash
No need to run `manage.py compilemessages`.
+ scripts/lib/clean-unused-caches
Cleaning unused venv caches...
Cleaning unused node modules caches...
Cleaning unused emoji caches...
writing to /srv/zulip/var/54db07ed-532f-40bb-acbe-1983e2d30eb9/provision_version
```

Reply:
We want them to run on the first two in `provision`, and on the second one in `test-backend`.

Reply:
yeah we don't need to run it on the third one since that is practically non-existent if it is copied from the template at the time of running each test. so if the template is up-to-date, tests run on latest schema

Reply:
But what I am talking about is different. I am saying that even after running it on the template db it seems like it says no migrations to apply. I am still trying to understand why?

Reply:
Idunno; it'll be a lot easier for me to help after you clean this branch up as I suggested, since the changes are hard to read.  But I'm guessing it's ultimately an issue of passing the wrong options.

Reply:
ohk, let me clean it up, first.

Reply:
Oh, the bug is that you probably shouldn't be using `call_command` for running migrations.  I think that's inheriting the already-imported Django settings and thus running against the dev database.  Shell out to `manage.py`.

Reply:
I actually thought of that as well. So what I did was change the environment variable `DJANGO_SETTINGS_MODULE` to zproject.test_settings, and do django.setup again before calling the run_db_migrations just as a way to check if that was it. Didn't work!

My next plan of action to check out what was wrong was to try using manage.py only :)

Reply:
I still suspect doing that alone wouldn't work since ./manage migrate doesn't work with --settings so probably we need to export DJANGO_SETTINGS_MODULE correctly